codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,467 @@
+ * Channel#read()}, which follows the default behavior of channels in Netty. Applications are
+ * free to close the channel in response to such events if they don't have use for any queued
+ * messages.
+ *
+ * <p>{@link ChannelConfig#setMaxMessagesPerRead(int)} and {@link
+ * ChannelConfig#setAutoRead(boolean)} are supported.
+ */
+public class Http2MultiplexCodec extends ChannelDuplexHandler {
+    private static final Http2FrameLogger HTTP2_FRAME_LOGGER
+        = new Http2FrameLogger(INFO, Http2MultiplexCodec.class);
+
+    private final ChannelHandler streamHandler;
+    private final EventLoopGroup streamGroup;
+    private final Http2ConnectionHandler http2Handler;
+    private final Http2Connection.PropertyKey streamInfoKey;
+    private final List<StreamInfo> streamsToFireChildReadComplete = new ArrayList<StreamInfo>();
+    private ChannelHandlerContext ctx;
+    private ChannelHandlerContext http2HandlerCtx;
+
+    /**
+     * Construct a new handler whose child channels run in the same event loop as this handler.
+     *
+     * @param server {@code true} this is a server
+     * @param streamHandler the handler added to channels for remotely-created streams. It must be
+     *     {@link ChannelHandler.Sharable}.
+     */
+    public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+        this(server, streamHandler, null);
+    }
+
+    /**
+     * Construct a new handler whose child channels run in a different event loop.
+     *
+     * @param server {@code true} this is a server
+     * @param streamHandler the handler added to channels for remotely-created streams. It must be
+     *     {@link ChannelHandler.Sharable}.
+     * @param streamGroup event loop for registering child channels
+     */
+    public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler,
+            EventLoopGroup streamGroup) {
+        this(server, streamHandler, streamGroup, new DefaultHttp2FrameWriter());
+    }
+
+    Http2MultiplexCodec(boolean server, ChannelHandler streamHandler,
+            EventLoopGroup streamGroup, Http2FrameWriter frameWriter) {
+        if (!streamHandler.getClass().isAnnotationPresent(Sharable.class)) {
+            throw new IllegalArgumentException("streamHandler must be Sharable");
+        }
+        this.streamHandler = streamHandler;
+        this.streamGroup = streamGroup;
+        Http2Connection connection = new DefaultHttp2Connection(server);
+        frameWriter = new Http2OutboundFrameLogger(frameWriter, HTTP2_FRAME_LOGGER);
+        Http2ConnectionEncoder encoder = new DefaultHttp2ConnectionEncoder(connection, frameWriter);
+        Http2FrameReader reader = new Http2InboundFrameLogger(new DefaultHttp2FrameReader(), HTTP2_FRAME_LOGGER);
+        Http2ConnectionDecoder decoder = new DefaultHttp2ConnectionDecoder(connection, encoder, reader);
+        decoder.frameListener(new FrameListener());
```

Comment:
Understood. The usage/scope of the escaping this in this scenario seems to be contained to this class. However the relationship is not obvious at the spot at which `this` escapes. It seems like this means an observer has to have a complete understanding of the entire scope of this class to know what is initialized and when. For example if a change is made to an inner class you may have to look outside the scope of that class to see what is initialized.

I'm fine with leaving as is because it is relatively self contained, but just something that we have to be mindful of in the future.